src/pkcs15init/pkcs15-openpgp.c

```diff
@@ -443,29 +443,48 @@ static int openpgp_store_data(struct sc_pkcs15_card *p15card, struct sc_profile
 	case SC_PKCS15_TYPE_CERT:
 		cinfo = (sc_pkcs15_cert_info_t *) obj->data;
 		cid = &(cinfo->id);
+		unsigned int tag = 0x7F21;
 
 		if (cid->len != 1) {
 			sc_log(card->ctx, "ID=%s is not valid.", sc_dump_hex(cid->value, cid->len));
 			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
 		}
 
 		/* OpenPGP card v.2 contains only 1 certificate */
-		if (cid->value[0] != 3) {
+		if (cid->value[0] != 3 && p15card->card->type < SC_CARD_TYPE_OPENPGP_V3) {
 			sc_log(card->ctx,
 			       "This version does not support certificate ID = %d (only ID=3 is supported).",
 			       cid->value[0]);
 			LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
 		}
-		/* Just update the certificate DO */
+
+		/* OpenPGP card < v.3 does not support SELECT DATA calls */
+		if (p15card->card->type >= SC_CARD_TYPE_OPENPGP_V3) {
+			/* Mapping [3..1] passed --id to [0..2] for param */
+			u8 param = (u8) (2 - (cid->value[0] - 1));
+			/* check for unsigned underflow */
+			if (param > 2) {
+				LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+			}
+
+			/* Just update the certificate DO */
+			r = sc_card_ctl(card, SC_CARDCTL_OPENPGP_SELECT_DATA, &param);
+			if (r < 0) {
+				LOG_TEST_RET(card->ctx, r, "Failed OpenPGP - select data");
+				LOG_FUNC_RETURN(card->ctx, r);
+			}
```

Comment:
this looks at least suspicious. First you check for error, then you check again and return on error (=always) and return and then you  return anyway. I think these three lines are functionally equivalent to the second one, if I see right:
```suggestion
			LOG_TEST_RET(card->ctx, r, "Failed OpenPGP - select data");
```

Reply:
done